### PR TITLE
Yet more compiler framework changes.

### PR DESCRIPTION
**Pull request by ben-clayton**

### CODE REVIEW

File: gapil/bapi/bapi.proto [hunk 1]
```diff
@@ -19,77 +19,163 @@ option go_package = "github.com/google/gapid/gapil/bapi";
...
+  repeated Choice choice = 18;
+  repeated Class class = 19;
+  repeated ClassInitializer class_init = 20;
+  repeated Clone clone = 21;
+  repeated Copy copy = 22;
+  repeated Create create = 23;
+  repeated DeclareLocal declare_local = 24;
+  repeated Definition definition = 25;
+  repeated DefinitionUsage definition_usage = 26;
   repeated Enum enum
```
Comment (AWoloszyn) [inline gapil/bapi/bapi.proto]: What is going on here? Is it because we have an enum named enum?

File: gapil/bapi/bapi.proto [hunk 2]
```diff
@@ -490,33 +610,92 @@ message Member {
...
+    uint64 cast = 18;           // Cast ID
+    uint64 choice = 19;         // Choice ID
+    uint64 class = 20;          // Class ID
+    uint64 class_init = 21;     // ClassInitializer ID
+    uint64 clone = 22;          // Clone ID
+    uint64 copy = 23;           // Copy ID
+    uint64 create = 24;         // Create ID
+    uint64 declare_local = 25;  // DeclareLocal ID
+    uint64 definition = 26;     // Definition ID
     uint64 enum
```
Comment (AWoloszyn) [inline gapil/bapi/bapi.proto]: Looks like it :\

File: gapil/bapi/bapi.proto [hunk 3]
```diff
@@ -695,199 +894,324 @@ message Write {
...
+    uint64 bool = 7;            // ASTBool
+    uint64 branch = 8;          // ASTBranch
+    uint64 call = 9;            // ASTCall
+    uint64 case = 10;           // ASTCase
+    uint64 class = 11;          // ASTClass
+    uint64 declare_local = 12;  // ASTDeclareLocal
+    uint64 default = 13;        // ASTDefault
+    uint64 definition = 14;     // ASTDefinition
+    uint64 delete = 15;         // ASTDelete
+    uint64 enum
```
Comment (AWoloszyn) [inline gapil/bapi/bapi.proto]: Here too

File: gapil/bapi/bapi.proto (hunk 1 above)
Comment (ben-clayton) [reply gapil/bapi/bapi.proto]: Yup, it is. I could rename it if you want - but would like to do so in a follow up CL.

File: gapil/compiler/types.go [hunk 4]
```diff
@@ -129,119 +131,132 @@ func (c *C) declareTypes() {
...
+				copyToTarget := c.M.Function(c.T.Void, "S_"+t.Name()+"_copy_to_target", c.T.CtxPtr, captureTypePtr, targetTypePtr).
+					LinkOnceODR().
+					Inline()
+
+				c.T.captureToTarget[t] = copyToTarget
+				c.Build(copyToTarget, func(s *S) {
+					src := s.Parameter(1).SetName("src")
+					dst := s.Parameter(2).SetName("dst")
+					for _, f := range t.Fields {
+						firstElem := src.Index(0, f.Name()).LoadUnaligned()
```
Comment (AWoloszyn) [inline gapil/compiler/types.go]: I don't love super deep nesting, but I don't actually have a solution for you right now anyway.

Comment (ben-clayton) [reply gapil/compiler/types.go]: Ack. I'll see if I can clean it up when I visit this code again.

**Pull request merged**